lib/puppet/provider/package/apt.rb

```diff
@@ -44,12 +47,43 @@ def checkforcdrom
     end
   end
 
+  def best_version(should_range)
+    available_versions = SortedSet.new
+
+    output = aptcache :madison, @resource[:name]
+    output.each_line do |line|
+      is = line.split('|')[1].strip
+      begin
+        is_version = DebianVersion.parse(is)
+        available_versions << is_version
+      rescue DebianVersion::ValidationFailure
+        Puppet.warning("Cannot parse #{is} as a debian version")
+      end
+    end
+
+    included_available_version = available_versions.select do |version|
+      should_range.include?(version)
+    end
+
+    return included_available_version[-1] unless included_available_version.empty?
+
+    Puppet.warning("No available version for package #{@resource[:name]} is included in range #{should_range}")
+    should_range
+  end
+
   # Install a package using 'apt-get'.  This function needs to support
   # installing a specific version.
   def install
     self.run_preseed if @resource[:responsefile]
     should = @resource[:ensure]
 
+    begin
+      should_range = VersionRange.parse(should, DebianVersion)
```

Comment:
maybe we should do this only if `should` is a String... I'm not sure about apt but in other providers symbols might get here as well